### PR TITLE
HPCC-32691 Add optional timestamp arg to createClientSpan()

### PR DESCRIPTION
**Pull request by mckellyln**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mckellyln** [comment]
@rpastrana here are my quick changes to be able to call createClientSpan() with a starting timestamp so that we can time something and if that time exceeds a threshold we then create the span.

**Comment by mckellyln** [comment]
I am using this code via -
```
                SpanTimeStamp SpanTS;
                SpanTS.setMSTickTime(startms);
                Owned<ISpan> addrCleanSpan = ctx->queryContextLogger().queryActiveSpan()->createClientSpan("method-name", &SpanTS);
                StringBuffer endPoint(host);
                endPoint.appendf(":%d", port);
                addrCleanSpan->setSpanAttribute("server", endPoint.str());
                addrCleanSpan->setSpanStatusSuccess(true);
```

**Comment by mckellyln** [comment]
@rpastrana - Should I create a real Pull Request for this ?

File: system/jlib/jtrace.cpp [hunk 1]
```diff
@@ -557,7 +557,7 @@ class CSpan : public CInterfaceOf<ISpan>
         return spanID.get();
     }
 
-    ISpan * createClientSpan(const char * name) override;
+    ISpan * createClientSpan(const char * name, const SpanTimeStamp * spanStartTimeStamp = nullptr) override;
     ISpan * createInternalSpan(const char * name) override;
```
Comment (rpastrana) [inline system/jlib/jtrace.cpp:561]: add optional spansparttime to createInternalSpan as well?

File: system/jlib/jtrace.cpp [hunk 2]
```diff
@@ -1001,9 +1001,14 @@ class CInternalSpan : public CChildSpan
 class CClientSpan : public CChildSpan
 {
 public:
-    CClientSpan(const char * spanName, CSpan * parent)
+    CClientSpan(const char * spanName, CSpan * parent, const SpanTimeStamp * spanStartTimeStamp = nullptr)
     : CChildSpan(spanName, parent)
     {
+        if (spanStartTimeStamp && spanStartTimeStamp->isInitialized())
```
Comment (rpastrana) [inline system/jlib/jtrace.cpp:1007]: if we add the starttime option to internal, it might make sense to move this to CChildSpan?

**Review by rpastrana — COMMENTED** [review]
@mckellyln seems fine, consider extending feature to createInternal. Also, should add unittets: https://github.com/hpcc-systems/HPCC-Platform/blob/9505b2defa7a85ab5e3f8a459bd620390d77e007/testing/unittests/jlibtests.cpp#L48

**Comment by mckellyln** [comment]
@rpastrana I've updated PR and added a test case.

File: testing/unittests/jlibtests.cpp [hunk 3]
```diff
@@ -482,11 +483,33 @@ class JlibTraceTest : public CppUnit::TestFixture
                 CPPUNIT_ASSERT_MESSAGE("Mismatched localParentSpanID detected",
                  strsame(serverSpanID, retrievedSpanCtxAttributes->queryProp("localParentSpanID")));
 
-                CPPUNIT_ASSERT_EQUAL_MESSAGE("Unexpected missing remoteParentID detected", true,
-                 retrievedSpanCtxAttributes->hasProp("remoteParentID"));
+                CPPUNIT_ASSERT_EQUAL_MESSAGE("Unexpected GlobalID detected", false,
```
Comment (rpastrana) [inline testing/unittests/jlibtests.cpp:486]: some of these changes seem unrelated.

File: testing/unittests/jlibtests.cpp [hunk 4]
```diff
@@ -482,11 +483,33 @@ class JlibTraceTest : public CppUnit::TestFixture
...
+                CPPUNIT_ASSERT_EQUAL_MESSAGE("Unexpected empty TraceID detected", false, isEmptyString(retrievedSpanCtxAttributes->queryProp("traceID")));
+                CPPUNIT_ASSERT_EQUAL_MESSAGE("Unexpected empty SpanID detected", false, isEmptyString(retrievedSpanCtxAttributes->queryProp("spanID")));
+            }
+        }
+
+        {
+            unsigned startMs = msTick() - 1234;
+            SpanTimeStamp SpanTS;
+            SpanTS.setMSTickTime(startMs);
+            OwnedSpanScope clientSpan = serverSpan->createClientSpan("clientSpan", &SpanTS);
```
Comment (rpastrana) [inline testing/unittests/jlibtests.cpp:503]: L:503 exercises the new constructor signature, but the bulk of this test seems unrelated to the functionality we want to test which is the span's manually declared start time. I think manualTestsDeclaredSpanStartTime could be used as a template to test this new functionality

**Review by rpastrana — COMMENTED** [review]
@mckellyln looks good, but a couple of questions re: unittests

**Comment by mckellyln** [comment]
I am not sure how to get unittest output into OTEL collector

File: testing/unittests/jlibtests.cpp [hunk 5]
```diff
@@ -231,6 +231,11 @@ class JlibTraceTest : public CppUnit::TestFixture
             OwnedSpanScope serverSpan = queryTraceManager().createServerSpan("declaredSpanStartTime", emptyMockHTTPHeaders, &declaredSpanStartTime);
             //{ "type": "span", "name": "declaredSpanStartTime", "trace_id": "0a2eff24e1996540056745aaeb2f5824", "span_id": "46d0faf8b4da893e",
             //"start": 1702672311203213259, "duration": 125311051 }
+
+            SpanTimeStamp clientSpanTimeStamp;
```
Comment (rpastrana) [inline testing/unittests/jlibtests.cpp:235]: were you able to confirm the start time was adjusted as expected?
We're not able to automatically validate these values (yet) but the log output should match your expected start time (we might need to printout clientSpanTimeStamp.

**Review by rpastrana — APPROVED** [review]
@mckellyln looks good, but I'm left wondering if there's enough in the manual test to confirm the feature worked as expected.

**Comment by rpastrana** [comment]
> I am not sure how to get unittest output into OTEL collector

Right now we're dependent on visual inspection of the log output

**Comment by mckellyln** [comment]
Added jlog output as comments to help verify output is expected.
Squashed.
Ready for merge.

**Review by rpastrana — APPROVED** [review]
@mckellyln approved

File: testing/unittests/jlibtests.cpp (hunk 5 above)
Comment (mckellyln) [reply testing/unittests/jlibtests.cpp:235]: yes, I added the comments for the jlog Span output

**Comment by mckellyln** [comment]
Thanks so much.
Ready for merge.
